enterprise/reporting/error-index/worker_test.go

```diff
@@ -0,0 +1,682 @@
+			}
+			require.EqualValues(t, []payload{piiPayload}, expectedPayloads)
+		})
+		t.Run("limits reached but few left without crossing upload frequency", func(t *testing.T) {
+			receivedAt, failedAt := time.Date(2021, 1, 1, 1, 1, 1, 0, time.UTC), time.Date(2021, 1, 1, 1, 1, 1, 0, time.UTC)
+
+			postgresContainer, err := resource.SetupPostgres(pool, t)
+			require.NoError(t, err)
+			minioResource, err := destination.SetupMINIO(pool, t)
+			require.NoError(t, err)
+
+			eventsLimit := 24
+
+			c := config.New()
+			c.Set("INSTANCE_ID", instanceID)
+			c.Set("Reporting.errorIndexReporting.minWorkerSleep", "1s")
+			c.Set("Reporting.errorIndexReporting.uploadFrequency", "600s")
+			c.Set("Reporting.errorIndexReporting.eventsLimit", strconv.Itoa(eventsLimit))
+
+			errIndexDB := jobsdb.NewForReadWrite("err_idx", jobsdb.WithDBHandle(postgresContainer.DB), jobsdb.WithConfig(c))
+			require.NoError(t, errIndexDB.Start())
+			defer errIndexDB.TearDown()
+
+			count := 100
+			eventPayloads := make([]payload, 0, count)
+			jobsList := make([]*jobsdb.JobT, 0, count)
+
+			for i := 0; i < count; i++ {
+				eventPayload := prepareEventPayload(i, sourceID, receivedAt, failedAt.Add(time.Duration(i)*time.Second))
+				eventPayloads = append(eventPayloads, eventPayload)
+
+				epJSON, err := json.Marshal(eventPayload)
+				require.NoError(t, err)
+
+				jobsList = append(jobsList, &jobsdb.JobT{
+					UUID:         uuid.New(),
+					Parameters:   []byte(`{"source_id":"` + sourceID + `","workspaceId":"` + workspaceID + `"}`),
+					EventPayload: epJSON,
+					EventCount:   1,
+					WorkspaceId:  workspaceID,
+				})
+			}
+			require.NoError(t, errIndexDB.Store(ctx, jobsList))
+
+			cs := newMockConfigFetcher()
+			cs.addWorkspaceIDForSourceID(sourceID, workspaceID)
+			cs.addPIIReportingSettings(workspaceID, false)
+
+			statsStore := memstats.New()
+
+			fm, err := filemanager.New(&filemanager.Settings{
+				Provider: warehouseutils.MINIO,
+				Config: map[string]any{
+					"bucketName":      minioResource.BucketName,
+					"accessKeyID":     minioResource.AccessKey,
+					"secretAccessKey": minioResource.SecretKey,
+					"endPoint":        minioResource.Endpoint,
+				},
+			})
+			require.NoError(t, err)
+
+			w := newWorker(sourceID, c, logger.NOP, statsStore, errIndexDB, cs, fm)
+			defer w.Stop()
+
+			for i := 0; i < count/eventsLimit; i++ {
+				require.True(t, w.Work())
+			}
+			require.False(t, w.Work())
+
+			jr, err := errIndexDB.GetUnprocessed(ctx, jobsdb.GetQueryParams{
+				ParameterFilters: []jobsdb.ParameterFilterT{
+					{Name: "source_id", Value: w.sourceID},
+				},
+				PayloadSizeLimit: w.config.payloadLimit.Load(),
+				EventsLimit:      int(w.config.eventsLimit.Load()),
+				JobsLimit:        int(w.config.eventsLimit.Load()),
+			})
+			require.NoError(t, err)
+			require.Len(t, jr.Jobs, 4)
+		})
+	})
+}
+
+func failedMessages(t testing.TB, ctx context.Context, mr *destination.MINIOResource, filePath string) []payload {
```

Comment:
Using duckdb here is a good idea, but using [minio select](https://github.com/minio/minio/blob/master/docs/select/README.md) would be a more accurate way to test the sources' behaviour. 

We can keep both tests, I would suggest reading the messages with s3 select as well

Reply:
Thanks for pointing it out. Incorporating minio select as well.

Reply:
Added it [here](https://github.com/rudderlabs/rudder-server/pull/4002/files#diff-4d36be3b50240ec080299c6e00339b5506027d5cfb2c0ce98524cbcb8ef1607cR494-R548).